app/src/main/java/org/greenstand/android/TreeTracker/messages/ChatScreen.kt

```diff
@@ -0,0 +1,291 @@
+                    },
+                    keyboardActions = KeyboardActions(
+                    onGo = {
+                        viewModel.sendMessage()
+                    }
+                ),
+                colors = TextFieldDefaults.textFieldColors(
+                    textColor = AppColors.LightGray,
+                    backgroundColor = AppColors.DeepGray,
+                    focusedIndicatorColor = Color.Transparent,
+                    unfocusedIndicatorColor = Color.Transparent,
+                    disabledIndicatorColor = Color.Transparent
+                ),
+                    maxLines = 2
+                )
+            }
+    }
+}
+}
+
+const val ConversationTestTag = "ConversationTestTag"
+
+@Composable
+fun Messages(
+    state: ChatState,
+    scrollState: LazyListState,
+    modifier: Modifier = Modifier,
+    viewModel: ChatViewModel
+) {
+    val messages = state.messages
+    Box(modifier = modifier) {
+        LazyColumn(
+            reverseLayout = true,
+            state = scrollState,
+            contentPadding = PaddingValues(top = 10.dp),
+            // Add content padding so that the content can be scrolled (y-axis)
+            // below the status bar + app bar
+            // TODO: Get height from somewhere
+            modifier = Modifier
+                .testTag(ConversationTestTag)
+                .fillMaxSize()
+        ) {
+
+            for (index in messages.indices) {
+                item {
+                    Message(
+                        msg = messages[index],
+                        isAdmin = viewModel.checkIsAdmin(index),
+                        isFirstMessageByAuthor = viewModel.checkChatAuthor(index,true),
+                        isLastMessageByAuthor = viewModel.checkChatAuthor(index,true)
+                    )
+                }
+            }
+        }
+    }
+}
+
+@Composable
+fun Message(
+    msg: DirectMessage,
+    isAdmin: Boolean,
+    isFirstMessageByAuthor: Boolean,
+    isLastMessageByAuthor: Boolean
+) {
+    val spaceBetweenAuthors = if (isLastMessageByAuthor) Modifier.padding(top = 8.dp) else Modifier
+    Row(modifier = spaceBetweenAuthors) {
+        TextMessage(
+            msg = msg,
+            isAdmin = isAdmin,
+            isFirstMessageByAuthor = isFirstMessageByAuthor,
+            modifier = Modifier
+                .padding(end = 16.dp)
+                .weight(1f)
+        )
+    }
+}
+
+@Composable
+fun TextMessage(
+    msg: DirectMessage,
+    isAdmin: Boolean,
+    isFirstMessageByAuthor: Boolean,
+    modifier: Modifier = Modifier
+) {
+    Column(modifier = modifier) {
+        ChatItemBubble(msg, isAdmin)
+        if (isFirstMessageByAuthor) {
+            // Last bubble before next author
+            Spacer(modifier = Modifier.height(8.dp))
+        } else {
+            // Between bubbles
+            Spacer(modifier = Modifier.height(4.dp))
+        }
+    }
+}
+
+@Composable
+fun ChatItemBubble(
+    message: DirectMessage,
```

Comment:
@Elforama yes I followed it. It's better this way. For the announcement message where we need links it won't be in the body it would be videoLink if I am correct. So instead of passing different parameters of DirectMessage we would just pass DirectMessage. Another use case is if we want to add time or date to the chat bubble.